### PR TITLE
Nick issue 9

### PR DESCRIPTION
**Pull request by ski-climb**
@lucyconklin @CPowell23 @Robbie-Smith, this should take care of item 9 and item 10, er user stories 9 and 10.  I feel like the groundwork has been laid so that story 12 (checking out) should be pretty reasonable to complete.  I might be in the best place to tackle that tomorrow - though it might be good for someone to pair with me on it in order to spread around familiarity with how Orders work.  Some of our assumptions (mine very much included) about the database structure didn't really play out as we anticipated.  

Oh, I also had to revamp our seeds file and work some factories magic too.  Please take a look and let me know what you think.  @Robbie-Smith and @CPowell23 I really like how you two have been breaking down your feature tests - I feel like mine might be trying to do too much and could use your feedback to get leaner and meaner (happier?).

Cheers!

### CODE REVIEW

File: spec/features/customers/orders/show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+describe "Viewing one of a customer's order" do
+  context "when a customer has an order" do
+    scenario "displays the detail page for the order" do
+      order_1, order_2, order_3 = create_list(:all_new_order, 3)
```
Comment (ski-climb) [inline spec/features/customers/orders/show_spec.rb]: Yes, Hound, we're making sure that order_1 doesn't appear on the page.  Oh, Hound, you dog.

File: app/models/order.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  def total_price_in_dollars
+    total_price_in_cents / 100.to_f
+  end
+
+  def date
+    created_at.strftime("%e %b %Y")
+  end
+
+  def items_and_quantities
+    items.inject(Hash.new(0)) do |hash, item|
```
[the remark this answers is not in the document]
Comment (lucyconklin) [reply app/models/order.rb]: Lets change this to .reduce

File: app/models/item.rb [hunk 3]
```diff
@@ -7,9 +7,11 @@ class Item < ApplicationRecord
   validates :retired, inclusion: { in: [true, false] }
 
   belongs_to :category
+  has_many :order_items
+  has_many :orders, through: :order_items
```
Comment (lucyconklin) [inline app/models/item.rb:12]: refactor: lets put price_in_dollars in a helper

File: db/migrate/20170107034803_create_orders.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class CreateOrders < ActiveRecord::Migration[5.0]
+  def change
+    create_table :orders do |t|
+      t.integer :total_price_in_cents
+      t.references :status, index: true, foreign_key: true
```
Comment (lucyconklin) [inline db/migrate/20170107034803_create_orders.rb:5]: Where is this calculated? It does not seem to be working as expected.

Comment (lucyconklin) [reply db/migrate/20170107034803_create_orders.rb:5]: re :total_price_in_cents

Comment (lucyconklin) [reply db/migrate/20170107034803_create_orders.rb:5]: @ski-climb

Comment (lucyconklin) [reply db/migrate/20170107034803_create_orders.rb:5]: We will need to add this to order_spec , order_index , and order_show

File: db/seeds.rb [hunk 5]
```diff
@@ -1,16 +1,53 @@
...
-  category = Category.create!(name: Faker::Commerce.department)
-  9.times do
-    category.items.create!( title: Faker::Commerce.product_name,
-                            description: Faker::Hipster.sentence,
-                            price_in_cents: Faker::Number.number(4),
-                            image_url: images[rand(3)])
+puts "Making some new items for your database... \n\n" 
+FactoryGirl.create_list(:category, 7)
+categories = Category.all
+number_from_10_to_30 = rand(20) + 20
```
Comment (lucyconklin) [inline db/seeds.rb]: This is a blatent lie

Comment (lucyconklin) [reply db/seeds.rb]: Faker::Number.between(1, 10)

File: spec/factories.rb [hunk 6]
```diff
@@ -1,9 +1,50 @@
...
+    name { ["paid", "cancelled", "completed", "ordered"].sample }
+  end
+
+  factory :order do
+    total_price_in_cents { Faker::Number.between(9_99, 500_01) }
+    customer { Customer.all.sample(1).first }
+    status { Status.all.sample(1).first }
+    items { FactoryGirl.create_list(:item, 5) }
+  end
+
```
Comment (lucyconklin) [inline spec/factories.rb:13]: could possibly be solved by adding a trait

File: spec/features/customers/orders/show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        expect(page).to have_content "#{quantity} x $#{item.price_in_dollars}"
+        expect(page).to have_content "$#{item.price_in_dollars(quantity)}"
+      end
+
+      expect(page).to have_content "Status: #{order_3.status_name}"
+      expect(page).to have_content "$#{order_3.total_price_in_dollars}"
+      expect(page).to have_content "submitted at: #{order_3.display_submitted_at}"
+    end
+  end
+end
```
Comment (lucyconklin) [inline spec/features/customers/orders/show_spec.rb:27]: At some point in this test we should check to make sure our order status cancelled or completed is behaving as expected (showing the partial)

File: spec/models/item_spec.rb [hunk 8]
```diff
@@ -15,14 +15,20 @@
...
+    it "returns the price in dollars" do
       expect(item.price_in_dollars).to eq 12.34
     end
+
+    it "returns the price in dollars when there is more than one item" do
+      expect(item.price_in_dollars(10)).to eq 123.40
+    end
   end
 
   describe "#category_name" do
```
Comment (lucyconklin) [inline spec/models/item_spec.rb:34]: We should add a new test here (and on all model specs) that a model can be saved under happy path conditions. We are testing that it needs these attributes to be valid, but we also need to make sure that a valid item is actually valid.

File: spec/models/order_item_spec.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+describe OrderItem do
+  describe 'associations' do
+    it { is_expected.to belong_to(:order) }
+    it { is_expected.to belong_to(:item) }
+  end
+
+  describe "before saving" do
+    context "saves the item quantity and price" do
+      it "returns the price of the item in cents" do
```
Comment (lucyconklin) [inline spec/models/order_item_spec.rb:11]: @ski-climb This should just be price.

File: spec/models/order_spec.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    it { is_expected.to validate_presence_of(:total_price_in_cents) }
+  end
+
+  describe 'associations' do
+    it { is_expected.to have_many(:order_items) }
+    it { is_expected.to have_many(:items).through(:order_items) }
+    it { is_expected.to belong_to(:status) }
+    it { is_expected.to belong_to(:customer) }
+  end
+
```
Comment (lucyconklin) [inline spec/models/order_spec.rb:14]: This would be a great place for a test for total_price_in_cents

File: spec/factories.rb [hunk 11]
```diff
@@ -1,9 +1,50 @@
 FactoryGirl.define do
+
```
Comment (ski-climb) [inline spec/factories.rb]: I'm using white space for organizational clarity.

File: db/migrate/20170107034803_create_orders.rb (hunk 4 above)
Comment (ski-climb) [reply db/migrate/20170107034803_create_orders.rb:5]: Thanks for catching this!  I've added some model specs, and this information is set before the order object gets saved.

File: spec/models/item_spec.rb (hunk 8 above)
Comment (ski-climb) [reply spec/models/item_spec.rb:34]: @CPowell23 added those.  Thank you!

File: spec/models/order_spec.rb (hunk 10 above)
Comment (ski-climb) [reply spec/models/order_spec.rb:14]: Done!

File: app/models/order.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  def total_price_in_dollars
+    total_price_in_cents / 100.to_f
+  end
+
+  def date
+    created_at.strftime("%e %b %Y")
+  end
+
+  def items_and_quantities
+    items.reduce(Hash.new(0)) do |hash, item|
```
Comment (ski-climb) [inline app/models/order.rb]: Negative, Hound.  It's just what it's supposed to be.

File: spec/factories.rb (hunk 6 above)
Comment (Robbie-Smith) [inline spec/factories.rb]: @lucyconklin What could possibly be solved?

Comment (Robbie-Smith) [inline spec/factories.rb]: @CPowell23 Do you know what @lucyconklin was referring to?

**Comment by robbie-smith** [comment]
@ski-climb @lucyconklin @CPowell23 
Merged

**Pull request merged**